Second update for the digitalSTROM-Binding

MichaelOchel opened the pull request:
# new supports:
* binary-input-sensors
* room temperation control
* BL-KM200 and dS-iSens200

# optimize:
* request query building by dS-JSON-API
* session-token handling
* separtion of EventListener and EventHandler
* and some more

# others:
* add DsDeviceThingTypeProvider
* complete DsChannelTypeProvider to generate ChannelTypes through
constants
* extented dS-JSON-API
* bug-, typos-, Java-Doc-fixes and much more


Also-by: Matthias Siegele <digitalSTROM-ESH-binding@web.de>
Signed-off-by: Michael Ochel <michael.ochel@koeln.de>

kaikreuzer commented:
Wow, thanks, this is huge 😲 

@hmerk Would you volunteer for some review/testing?

hmerk commented:
Wow, I will try to find some time to review, testing might be difficult, as I do not have the newly introduced devices.
I will see what I can do.....

kaikreuzer commented:
Regression testing would also be helpful already, just to confirm that existing stuff does not break :-)

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/OSGI-INF/DigitalSTROMBridgeMDNSDiscovery.xml:
```diff
@@ -8,10 +8,13 @@
     http://www.eclipse.org/legal/epl-v10.html
 
 -->
+<!-- Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others. All rights reserved. This program and the accompanying 
```
Please use correct license information

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/OSGI-INF/DigitalSTROMBridgeMDNSDiscovery.xml:
```diff
@@ -8,10 +8,13 @@
 -->
+<!-- Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others. All rights reserved. This program and the accompanying 
+	materials are made available under the terms of the Eclipse Public License v1.0 which accompanies this distribution, and 
+	is available at http://www.eclipse.org/legal/epl-v10.html -->
```
Please use correct license information

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/OSGI-INF/DigitalSTROMChannelTypeProvider.xml:
```diff
@@ -8,6 +8,9 @@
 -->
+<!-- Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others. All rights reserved. This program and the accompanying 
+	materials are made available under the terms of the Eclipse Public License v1.0 which accompanies this distribution, and 
+	is available at http://www.eclipse.org/legal/epl-v10.html -->
```
Please use correct license information

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -15,71 +15,67 @@ The digitalSTROM-Server is required for accessing any other digitalSTROM-Devices
 
-At this point almost all available **GE**, **SW** and **GR** digitalSTROM-Devices with a set output-mode, unequal *disabled* or *PWM*, are supported by this binding.  
-For that there are identically named thing types. Only the *GR* type has a channel (shade), which cannot be changed. The other two types add their channels dynamically affected by the set color-group and output-mode. They also automatically change or add the channels, if the color-group or output-mode has changed through the dSS-web-configuration or the configured sensor priorities of the thing h ...
+At this point almost all available **GE**, **SW**, **GR** and **BL** digitalSTROM-Devices with an set output-mode, unequal *disabled*, are supported by this binding. Furthermore sensor devices like the **dS-iSens200** and **SW-devices** with binary-inputs are supported. Last but not least the **circuit** (dS-Meter) is supported, too. They will provide the power consumption and electric meter as c ...
```
Typo : a set output mode

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -15,71 +15,67 @@ The digitalSTROM-Server is required for accessing any other digitalSTROM-Devices
 
-The following table shows all tested digitalSTROM-Devices with their output-modes.
+For that there are identically named thing types. Only the *GR* type has a channel (shade), which cannot be changed. The other types adds their channels dynamically affected by the set color-group and output-mode. They also automatically change or add the channels, if the color-group or output-mode has changed through the dSS-web-configuration or the configured sensor priorities of the thing has  ...
+
```
Typo : The other types add their channels

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -15,71 +15,67 @@ The digitalSTROM-Server is required for accessing any other digitalSTROM-Devices
+| BL-KM200 | switch, pwm | blue |
+
+- Binary-inputs were tested with SW-UMR200.
+- Sensor channels were tested with dS-iSens200 and power sensor with all other supported devices, which are listed in the upper table.
```
which are listed in the above table

msiegele replied:
I think it has to be "listed in the table above".

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -15,71 +15,67 @@ The digitalSTROM-Server is required for accessing any other digitalSTROM-Devices
+### digitalSTROM-Zone-Temperature-Control
+
+Last but not least, the digitalSTROM-Zone-Temperature-Control is also supported, if a zone-temerature-control is configured, as thing-type **zone_temperature_control**. The difference between the digitalSTROM-heating-control-app is, that their are no operation-modes, like *comfort* or *eco*. You can directly set the target temperature, in case *pid-control* is configured, otherwise you can set th ...
+
```
Typo : that there are no operation modes

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -15,71 +15,67 @@ The digitalSTROM-Server is required for accessing any other digitalSTROM-Devices
-2. restart the server to delete the temporary saved scene configuration 
-3. add the thing again to read out the scene configuration again.   
+If you only want to discover one of the thing types, you can start a discovery scan on the thing type you wish have discovered. You can use the command line command, e.g.: ``smarthome:discovery start digitalstrom:namedScene`` to start the scan. What thing types this binding supports please have a look at **Supported Things**. 
 
```
... on the thing type you want to have discovered.

For thing types this binding .....

msiegele replied:
We changed it to "Which thing types this binding supports, please ...

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -15,71 +15,67 @@ The digitalSTROM-Server is required for accessing any other digitalSTROM-Devices
 
-If the digitalSTROM-Server isn’t found automatically, e.g. because the server isn’t placed at the local network or the mDNS-service is deactivated, you have to insert the network address or URL and the authentication data manually through the graphical user interface or type it into the \*.thing with textual configuration.
+If the digitalSTROM-Server isn’t found automatically, e.g. because the server isn’t placed at the local network or the mDNS-service is deactivated, you have to insert the network address or URL and the authentication data manually through the graphical user interface or type it into the \*.thing with textual configuration. If you use for authentication the user name and password, the application- ...
 
```
If you use the user name and password for authentication ande there already is a token for this application, it will be automatically retreived from the digitalSTROM-Server, otherwise a new application-token will be generated.

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -109,88 +104,255 @@ Here you can set general binding configuration parameters, which are shown in fo
+#### Properties
+
+In contrast to the digitalSTROM-Device there are other informatically properties. The following table shows the available properties:
+
```
informal properties

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -105,61 +127,105 @@ public Initializer(BridgeHandler bridge, Config config) {
+                if (StringUtils.isBlank(getThing().getProperties().get(DigitalSTROMBindingConstants.SERVER_CERT))
+                        && StringUtils.isNotBlank(config.getCert())) {
+                    properties.put(DigitalSTROMBindingConstants.SERVER_CERT, config.getCert());
+                    // updateProperty(DigitalSTROMBindingConstants.SERVER_CERT, config.getCert());
```
remove this line

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -105,61 +127,105 @@ public Initializer(BridgeHandler bridge, Config config) {
 
+    // @Override
+    private void thingUpdatedInt(Thing thing) {
+        // dispose();
```
remove comment

hmerk commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -105,61 +127,105 @@ public Initializer(BridgeHandler bridge, Config config) {
+    private void thingUpdatedInt(Thing thing) {
+        // dispose();
+        this.thing = thing;
+        // initialize();
```
remove comment

msiegele commented:
Hello everyone,

we would like to ask how it looks with the pull request and the tests. 
Is there any progress or did you have an idea when it goes forward? 

Best regards 

Michael and Matthias

msiegele commented:
We wanted to solve the two merge conflicts ... maybe this was a bad idea.

sjka commented:
rebasing your branch would have been the better way... 😄

hmerk commented:
@msiegele 
sorry for being quiet for some time, but did not find time for further review.
The progress I made was using your updated version in my environment now.
It should be mentioned in the readme that this update needs actual firmware running on the DSS, otherwise, things won't come online.
I still had an old version with root access and had to update.
Apart from that, tests look really good with GE/GR/SW devices.

hmerk commented:
@msiegele 
Changes regarding I18n introduced with ESH b5 / openHAB 2.1 release broke your new version.
Could you please check and adapt your changes.

Best
Hans-Jörg

msiegele commented:
@hmerk 

Sorry for the late answer, but the last weeks were a bit busy.

We think you mean the renaming of the I18nProvider to TranslationProvider at #3615. So we think we have to rebase the branch to get the changes and solve the problem. 

But, the mistake by the attempt to solve the two merge conflicts directly through github and the resulting merge of the master into our branch, concludes to a lot of commits at the rebase process.

It looks like that:
cmd:
git rebase -i HEAD~3
result:
pick d9bc595f6 update-2
pick a406a0629 incorparation of the first part of CR
pick b12e19c3d WeMo Fix for 3183 - IllegalArgumentException on OSRAM Lightify Flex discovery (#3188)
pick 463e6601e Extensions search on label and id field (#3189)
pick b5fa767d6 Update the template for the BindingConstants file (#3198)
pick c6e9b59b2 add groovy DSL support to classpath (#3201)
… 165 more commits
pick d9b021e9e fixed FolderObserverTest for hidden files on windows systems (#3490)

Did we see it right, that we could squash all the commit, because they are merged into our branch and that they are not the commits of the master? If we can’t squash it, what is the best way to solve the problem or is there generally a better way (except directly using rebase and not merge)?
We tested it already (after a backup) at our local branch. We thought, that the local branch will be automatically updated, but it doesn’t seem like that. Did we do something wrong or will it only update after a forced push?

By the way, should we allready squash the current commits?

Best regards 

Michael and Matthias

triller-telekom commented:
Since you are now left with a fairly "strange looking" history of commits already because of the wrong merge commit I think the easiest solution for you will now be to checkout the latest master, create a new branch from there and then cherry-pick all of your commits onto it. Afterwards you can do a "force push" over your online branch "MichaelOchel:digitalSTROM-Binding_update-2-sq".

The result will then be the current history of ESH + all your commits on top of it. Once your PR is finished and approved all your commits will be squashed into one single commit and then applied on top of the current master by the merging person.

I hope this explanation helps. If not feel free to ask :)

msiegele commented:
Hi @triller-telekom ,

at first thank you for your answer.

We already created a new local branch from “eclipse/master” named “digitalSTROM-Binding_update-2-sq-fix-merge”. As next we “cherry-pick” the two commits with
“git cherry-pick d9bc595f6 
git cherry-pick a406a0629”
to the new local branch. 

Then we fixed the problem with a further commit by renaming the i18nProvider to TranslationProvider. So the history of our new local branch looks like that:

update-2
incorporation of the first part of CR
rename i18nProvider to TranslationProvider

We currently think it looks good that way, do you agree?

Next step would be the “force push”. Because we think, we can destroy a lot with this step here another question. Is the following command: `“git push MichaelOchel:digitalSTROM-Binding_update-2-sq digitalSTROM-Binding_update-2-sq-fix-merge -f”` correct?

And do we see it right, that we might have to do another rebase before or do the steps before again? Sometimes the branch is really fast outdated. E.g. our new local branch was already one commit behind “eclipse/master” after we finished the cherry-pick and commit steps in about an half hour. 

Best regards 

Michael and Matthias

triller-telekom commented:
Hello Michael and Matthias,

taking the current eclipse/smarthome master branch and then doing this pick:
> git cherry-pick d9bc595
> git cherry-pick a406a06

Sound right to me. But with this you should end up with at least this commit being the last before yours (as of now) `6e16762ef0be097be997719df861fe4941fdc3fe` which is `WU binding: fix a typo in a channel name in documentation (#3897)`. I do not know why you have this very old commit `rename i18nProvider to TranslationProvider` just below your 2 commit...

Are you sure you fetched the current master branch before creating your new local branch to pick your commits?

Basically what you need to do is:

1. git checkout master
2. git pull (this pulls the current origin/master into your local master)
3. git checkout -b newBranch
4. git cherry-pick a406a06
5. git cherry-pick d9bc595

Here your history should look like
* update-2
* incorporation of the first part of CR
* WU binding: fix a typo in a channel name in documentation (#3897) <- or newer!

If there are some conflicts due to the new commits you need to resolve them in another commit.

Afterwards you can force push this branch over your remote branch:

6. git push -f MichaelOchel/digitalSTROM-Binding_update-2-sq

where I am assuming your `git remote` is called "MichaelOchel".

MichaelOchel commented:
Hi  @triller-telekom 

oh, we think there was a misunderstanding. The commit “rename i18nProvider to TranslationProvider” is a next commit from us to solve the following problem.

> @msiegele
Changes regarding I18n introduced with ESH b5 / openHAB 2.1 release broke your new version.
Could you please check and adapt your changes.
> 
> Best
> 
>  Hans-Jörg

The correct history currently looks like that:

- d9baaffd75a8ce797bbfa484ed33309e08afc73c rename i18nProvider to TranslationProvider
- de8288a4b9b1da60cea0af9f8e7c6d9e445183ae incorparation of the first part of CR
- 1df1173ef71638e095d54900ca62f6e48744d965 update-2
- 1fb653aced1d58ccdd803bb50d856014e3b984f2 Fix Eclipse Oxygen Installer (#3866)
- …

But before we push, we will update the branch again to the latest eclipse/master.

So the next steps will be:
1. Update the branch again with rebase or do the procedure again.
2. “force push” with `git push -f MichaelOchel/digitalSTROM-Binding_update-2-sq`

Best regards 

Michael and Matthias

triller-telekom commented:
@MichaelOchel What is the status now? Did you manage to rebase it or do you need further help?

msiegele commented:
Hi  @triller-telekom 

sorry for the late answer.

We think we need further help. 

Our status is:

1. We tried to rebase the new created branch to get the latest master. Git says successful rebased, but the history was the same like before. So we created a further branch and do the cherry-pick procedure again.
2. Then we tried the ‘git push -f MichaelOchel/digitalSTROM-Binding_update-2-sq’ command. The result was the following exception: “fatal: The current branch digitalSTROM-Binding_update-2-sq-fix-merge2 has no upstream branch.To push the current branch and set the remote as upstream, usegit push --set-upstream MichaelOchel/digitalSTROM-Binding_update-2-sq digitalSTROM-Binding_update-2-sq-fix-merge2”
3. To test the proposed command from git we created a test branch and publish it. At this branch we wanted to test the “branch-override” first. So we do the both push commands for this branch. That resulted the following exception:
“fatal: 'MichaelOchel/test-override' does not appear to be a git repository
fatal: Could not read from remote repository.Please make sure you have the correct access rights
and the repository exists.”
The same exception was thrown directly by using the command “git push -f <remote> <branch>” with the new test branch as remote (MichaelOchel/test-override)  and the branch with the corrected history as branch. We think the exception is thrown, because MichaelOchel/digitalSTROM-Binding_update-2-sq respectively MichaelOchel/test-override is not add in ‘git remote’. The automatically added remotes are:
“git remote -v
eclipse https://github.com/eclipse/smarthome.git (fetch)
eclipse https://github.com/eclipse/smarthome.git (push)
origin  https://github.com/MichaelOchel/smarthome.git (fetch)
origin  https://github.com/MichaelOchel/smarthome.git (push)”
By the way “origin/<branch>” has the same effect.

So what do we have to do now? Add it as remote, use another command to push or something else?

Best regards 

Michael and Matthias

triller-telekom commented:
Branches do not show up in the remotes, you only have repositories there.

For the test scenario you did not post which commands you entered and so it is really hard to follow what you did. At least I do not know what you did by reading just the console outputs and not the corresponding commands.

I was also slightly wrong on the syntax of my push command, sorry...

Reading from your `git remote` output you have to push the newly created and rebased branch (lets call it newBranch) with this command:

`git push -f origin newBranch:digitalSTROM-Binding_update-2-sq`

because `origin` is YOUR repository on github and that is where the branch `digitalSTROM-Binding_update-2-sq` exists. And this command will replace the content of the remote branch `digitalSTROM-Binding_update-2-sq` with the content of your local `newBranch`.

MichaelOchel commented:
@triller-telekom 
this command syntax has worked now. We hope that it is all correct now. 
Thank you very much for the help.

And next time we will write the input, too. ;-)


@hmerk 
Now the binding should work again with the latest openHAB2 server.


Best regards 

Michael and Matthias

hmerk commented:
Thanks Michael and Mathias, and even thanks to Stefan helping with the git issue.
I will start testing again asap and keep you posted.

Best
Hans-Jörg

maggu2810 commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/manager/impl/StructureManagerImpl.java:
```diff
@@ -28,78 +30,95 @@
+            this.zoneID = zoneID;
+            this.zoneName = zoneName;
+
+            HashMap<Short, String> groupIdNames = new HashMap<Short, String>(groups.size());
```
As we are using Java 8 we can use the diamond operator that comes with Java 7

maggu2810 replied:
Any reason to first introdocue the local variable and later assign the local variable to the member variable?
Why not simplify this one by

```Java
groupIdNames = new HashMap<>(groups.size());
```

I like your size hint usage.

MichaelOchel replied:
I think I wasn't sure at this time, if I can put something on a final HashMap. But it works and thus `groupIdNames = new HashMap<>(groups.size());` is better. So you are right.

maggu2810 commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/manager/impl/StructureManagerImpl.java:
```diff
@@ -110,22 +129,21 @@ public boolean generateZoneGroupNames(ConnectionManager connectionManager) {
+        if (zoneGroupIdNameMap.get(zoneID) != null) {
+            if (zoneGroupIdNameMap.get(zoneID).groupIdNames != null) {
+                if (zoneGroupIdNameMap.get(zoneID).groupIdNames.get(groupID) != null) {
+                    return zoneGroupIdNameMap.get(zoneID).groupIdNames.get(groupID);
```
Any reason that we call the `get` method multiple times?

`zoneGroupIdNameMap.get(zoneID)` is called four times (one time should be enough)
`groupIdNames.get(groupID)` is called twice (onen time should be enough)

So, instead of call get and if it is non null call it again, it would be better to store the result to a local variable and reuse it.

MichaelOchel replied:
Currently I think all if-blocks can be deleted or rather reduce to one. 
Because of the new object, the HashMap for the group-names can't be null any more.

So the code can look like that:
`return zoneGroupIdNameMap.get(zoneID) != null ? zoneGroupIdNameMap.get(zoneID).groupIdNames.get(groupID) : null;`

Or with a local variable for `zoneGroupIdNameMap.get(zoneID)`. But I’m not really sure which way is more efficient.
The local variable has to be created every time. On the other side, two HashMap accesses are only, if it is not null.

And lines [176](https://github.com/eclipse/smarthome/pull/3182/files/a6b88cd6d4fc8225bb2f6e828ce380cd4c7b1c1b#diff-15ed87cab5c37f4262bc4152fa0ea898R176) - [181](https://github.com/eclipse/smarthome/pull/3182/files/a6b88cd6d4fc8225bb2f6e828ce380cd4c7b1c1b#diff-15ed87cab5c37f4262bc4152fa0ea898R181) can be changed to:

`ZoneGroupsNameAndIDMap zoneGroupsNameID = zoneGroupNameIdMap.get(zoneName);`
`        if (zoneGroupsNameID != null) {`
`            return zoneGroupNameIdMap.get(zoneName).groupNameIds.get(groupName) != null
                    ? zoneGroupNameIdMap.get(zoneName).groupNameIds.get(groupName)
                    : -1;`

maggu2810 replied:
Hm, I am not sure if I understand your doubts.

```Java
return map.get(key1) != null ? map.get(key1).get(key2) : null;
```

* if the first `map.get(key1)` is equal to `null` you have one `get`
* if the first `map.get(key1)` differs to `null` you have
** one additional `get` and a NPE if your key1 has been removed from the map by another thread (perhaps not possible in your impl. - I don't know). So 2x `get` + 1x NPE
** two additional `get`. So 3x `get`

```Java
final Map<...,...> tmp = map.get(key1);
return tmp != null ? tmp.get(key2) : null;
```

* one get if tmp equals to `null`
* one additional get if tmp differs to `null`
* you have one additional variable in the code, but this should be inlined by the byte compiler (AFAIK)

MichaelOchel replied:
I didn’t respect the case, that the first key can be removed by another thread. I think a NPE is improbable, but not impossible. So the local variable is safer and the better way. Thanks for the explanation.

In my opinion it is better when I add getter methods at the ZoneGroupsNameAndIDMap object to get the group-IDs and names. That do it more clearly that this maps can not be change.

I think it is useful when I incorporate the last comments etc. and commit it. That we get the same stand.

maggu2810 commented:
What is the state of this PR? If it is ready to review / merge, then please resolve the conflicts.

MichaelOchel commented:
We don’t really know what the state of the PR is. We wait for the review or other messages from e.g. @hmerk  about the functional test.

But I found another possible NPE trap at the StructureManager which I want to fix. But last weeks I didn’t find the time for that. I hope next week I have the time and can commit it.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/DeviceStateUpdate.java:
```diff
@@ -19,32 +21,37 @@
-    public final static String UPDATE_OUTPUT_CURRENT = "outputCurrent";
-    public final static String UPDATE_ELECTRIC_METER = "electricMeter";
     public final static String UPDATE_OUTPUT_VALUE = "outputValue";
+    public final static String UPDATE_DEVICE_SENSOR = "deviceSensor-";
```
Is the "-" at the end intended? Because I have seen places in the ode where you explicitly do a string concatenation with "variable - variable - variable"...

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/DeviceStateUpdate.java:
```diff
@@ -19,32 +21,37 @@
+    public final static String UPDATE_CIRCUIT_METER = "circuitMeter";
+
+    // binary inputs
+    public static final String BINARY_INPUT = "binaryInput-";
```
Is the "-" at the end intended? Because I have seen places in the ode where you explicitly do a string concatenation with "variable - variable - variable"...

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/ChangeableDeviceConfigEnum.java:
```diff
@@ -20,5 +20,6 @@
     FUNCTIONAL_GROUP,
-    OUTPUT_MODE;
+    OUTPUT_MODE,
+    BINARAY_INPUTS;
```
BINARY_INPUTS

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/DeviceParameterClassEnum.java:
```diff
@@ -60,7 +60,7 @@
      * @return index
      */
-    public int getClassIndex() {
+    public Integer getClassIndex() {
```
Why?

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/FuncNameAndColorGroupEnum.java:
```diff
@@ -5,20 +5,21 @@
+ *      "Table 1: digitalSTROM functional groups and their colors", page 9</a>
  */
-public enum FunctionalNameAndColorGroupEnum {
+
```
Please remove this empty line

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/FuncNameAndColorGroupEnum.java:
```diff
@@ -38,53 +39,53 @@
      * Returns true, if contains the given output mode id in DigitalSTROM, otherwise false.
      *
-     * @param functionalNameGroupID
+     * @param functionalNameGroupID to contains
```
to contains? what does this mean?

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/FunctionalColorGroupEnum.java:
```diff
@@ -40,23 +40,23 @@
-    WHITE(Lists.newArrayList(-1)),
-    RED(Lists.newArrayList(-2)),
-    GREEN(Lists.newArrayList(-3));
+    YELLOW(Lists.newArrayList((short) 1)),
```
Is it really important in terms of sending bits to the devices that you need shorts instead of ints here?

MichaelOchel replied:
That’s the ids we get from digitalSTROM and they are short formatted. We can use int too, but why, when we only need a short?

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/FunctionalColorGroupEnum.java:
```diff
@@ -65,33 +65,33 @@
      * Returns true, if contains the given functional color group id in digitalSTROM exits, otherwise false.
      *
-     * @param functionalColorGroupID
+     * @param functionalColorGroupID to contains
```
see above...

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/MeteringTypeEnum.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+import com.google.common.collect.Lists;
```
Please do not use the google library here

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/OutputModeEnum.java:
```diff
@@ -85,7 +95,7 @@
      * Returns true, if the output mode id contains in digitalSTROM, otherwise false.
      *
-     * @param modeID
+     * @param modeID to contains
```
see above, i dont know what this means

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/constants/SensorEnum.java:
```diff
@@ -0,0 +1,244 @@
+    private final Float max;
+    private final Float resolution;
+
+    static final HashMap<Short, SensorEnum> SENSOR_ENUMS = new HashMap<Short, SensorEnum>();
```
`static final Map<Short, SensorEnum> SENSOR_ENUMS = new HashMap<Short, SensorEnum>();`

Please also have a look at other places where you define lists or maps so you use the interface type in the definition and the specific one in the initialization.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/impl/DeviceBinaryInput.java:
```diff
@@ -0,0 +1,175 @@
+
+    private Logger logger = LoggerFactory.getLogger(DeviceBinaryInput.class);
+
+    private Short targetGroupType = null;
```
`null` is the default value and can be removed, please check at all places

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/impl/DeviceBinaryInput.java:
```diff
@@ -0,0 +1,175 @@
+                stateValue = jsonObject.get(JSONApiResponseKeysEnum.STATE.getKey()).getAsShort();
+            }
+        } catch (IllegalArgumentException e) {
+            logger.error("A IllegalArgumentException occurred: ", e);
```
This will end up in the users logfile who does not know what this is... Either change the severity to debug or if you want to let the user know so he can report the error, please give a description what the exception mean in this context. This also applies for the other exceptions you are catching here.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/impl/DeviceStateUpdateImpl.java:
```diff
@@ -0,0 +1,265 @@
+     * @param updateType must not be null
+     * @param value must not be null
+     */
+    public DeviceStateUpdateImpl(String updateType, Object value) {
```
This catches all other `public DeviceStateUpdateImpl(String updateType, XXX value)` constructors, no need for them.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/deviceParameters/impl/JSONCachedMeteringValueImpl.java:
```diff
@@ -0,0 +1,108 @@
+    private String date = null;
+    private MeteringTypeEnum meteringType = null;
+    private MeteringUnitsEnum meteringUnit = null;
+    private Logger logger = LoggerFactory.getLogger(CachedSensorValue.class);
```
wrong logger class

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/impl/DeviceImpl.java:
```diff
@@ -0,0 +1,1859 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
```
you probably do not need this import

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/impl/DeviceImpl.java:
```diff
@@ -0,0 +1,1859 @@
+                                sceneOutputMap.get(sceneNumber)[0]));
+                        if (isBlind()) {
+                            updateInternalDeviceState(new DeviceStateUpdateImpl(DeviceStateUpdate.SLAT_ANGLE,
+                                    sceneOutputMap.get(sceneNumber)[1]));
```
this looks like another "magic contract", who defines that position 1 of `sceneOutputMap.get(sceneNumber)` is responsible for blinds? A named reference would be more readable and maintainable...

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/impl/DeviceImpl.java:
```diff
@@ -0,0 +1,1859 @@
+    }
+
+    @Override
+    public boolean setDsSensorValue(SensorEnum sensorType, Integer dSSensorValue) {
```
Why are here 2 functions, one for type and one for index? I guess type and index are related in a sense that both should always be changed together and kept in sync? If that is the case I would clearly vote for creating only ONE method to set both... Same applies to the other setters here.

MichaelOchel replied:
No, some functions or events from digitalSTROM use or return the index and some other the type.

And the maping of the type to the index can be different by every device-type.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/scene/SceneDiscovery.java:
```diff
@@ -52,9 +52,10 @@
-    private final String reachableScenesQuery = "/json/zone/getReachableScenes?id=";
-    private final String reachableGroupsQuery = "/json/apartment/getReachableGroups?token=";
+    public final static String NAMEND_SCENE_QUERY = "/apartment/zones/*(ZoneID)/groups/*(group)/scenes/*(scene,name)";
+    // TODO: can be add to dSApi
```
What about this TODO?

MichaelOchel replied:
The same like the other TODO, no more time to implement it.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/scene/constants/ApartmentSceneEnum.java:
```diff
@@ -69,15 +71,15 @@ public static ApartmentSceneEnum getApartmentScene(short sceneNumber) {
      * Returns true, if the given scene number contains in digitalSTROM apartment scenes, otherwise false.
      *
-     * @param sceneNumber
+     * @param sceneNumber to contains
```
unclear what this means, see above

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/scene/constants/SceneEnum.java:
```diff
@@ -145,15 +148,15 @@ public static SceneEnum getScene(short sceneNumber) {
      * Returns true, if the given scene number contains in digitalSTROM scenes, otherwise false.
      *
-     * @param sceneNumber
+     * @param sceneNumber to contains
```
see above

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/scene/constants/ZoneSceneEnum.java:
```diff
@@ -48,15 +48,15 @@ public static ZoneSceneEnum getZoneScene(short sceneNumber) {
      * Returns true, if the given scene number contains in digitalSTROM zone scenes, otherwise false.
      *
-     * @param sceneNumber
+     * @param sceneNumber to contains
```
see above

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/providers/DsDeviceThingTypeProvider.java:
```diff
@@ -0,0 +1,139 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
```
you do not need this import here, lists can be created without

msiegele commented:
Hello @triller-telekom ,

first thanks for the extensive code review. This binding was written as a part of our bachelor thesis. This also explains why we could not put it into more PRs and by some parts there is truly a better concept possible, but within the short time this was not possible. 

Some of the points you had addressed in your review, for example the scene settings stored into the thing properties were discussed with @kaikreuzer . We know, that this is not the real idea behind the concept, but due to the special features of digitalSTROM, it was the best way. If you are interested in more documentation of this project, we can send you our bachelor thesis.

It will take some time to incorporate it, because we are working now or search for a job.

Our plan to incorporate your annotations, is that we will work off first the general parts, like remove default keywords and so on. After that we will incorporate the other things you have markt at the code. All of that, we will put it in much “smaller” commits. Unfortunately the general parts have very much changes. Between that we can discuss some things.

Best regards

Michael Ochel and Matthias Siegele

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -203,54 +365,170 @@ All devices support some of the following channels:
-    GE GE-KM200 [ dSID="3504175fe0000000000043d4",  ActivePowerRefreshPriority="low", ElectricMeterRefreshPriority="medium", OutputCurrentRefreshPriority="high"]
-    GR GR-KM200 [ dSID="3504175fe0000000000043d5"]
-    namedScene scene1 [zoneID="0", groupID="0", sceneID="5"]
+Bridge digitalstrom:dssBridge:dSS [ ipAddress="urlOfMyDss",  userName="dssadmin", password="mySecretPassword", sensorDataUpdateInterval=180] {
```
Both are possible, but we change "ipAddress" to "dSSAddress", because we think it is more specifical.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -203,54 +365,170 @@ All devices support some of the following channels:
+1. If a device status has changed through a device scene or a directly set output value, the status is maybe not correct.
+2. If the last called group scene does not read out yet, it can takes some time till the status will be updated. 
+
+### Textual configuration notice
```
That's because of the special scene-concept of digitalSTROM. And it's currently the best way we think.
Without this “feature”, all relevant scenes have to be read out after a server reboot and this is subject to a rule that says that "only 1 reading job per minute per circuit is allowed.". So that can be take much time.

It is a bit complicated to explain it complete.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/README.md:
```diff
@@ -203,54 +365,170 @@ All devices support some of the following channels:
+If you want to create a rule, which uses things of the digitalSTROM-Binding, there are also two things you have to  consider.
+
+1. If the rule contains several digitalSTROM-Devices, which can be summarized in a digitalSTROM-Scene, e.g. some lights in a zone, please use a equivalent supported scene. That will significantly reduce the comments to the digitalSTROM-Server, increases performance and does not bypasses the digitalSTROM state-machine.
+2. If you implement your own temperature control algorithm for a zone/room, e.g. because you want to use other temperature sensors, and call more than one digitalSTROM-BL-KM200, please use the *zone_temperature_control* for valve value control. The *zone_temperature_control* for valve value control will call all digitalSTROM-BL-KM200 with one command and increases so the performance. To get the n ...
```
There are more than one control mode and the user has to choose one of these, but NOT pid-controlled.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -104,61 +127,101 @@ public Initializer(BridgeHandler bridge, Config config) {
+                    updateConfiguration(configuration);
+                }
+                thingUpdatedInt(getThing());
+            } catch (Exception e) {
```
The ThreadPoolManager eats all exceptions, so it is really hard to find a bug, but we delete it. Maybe it has changed, we don’t know. The same explanation applies by all other try catch blocks in a run.

triller-telekom replied:
Which `ThreadPoolManager` is eating your exceptions here? The only purpose of your thread is to execute the content of your `run` method once and afterwards the thread is dead, no matter if it was successful or an exception was thrown. So I would not catch the exception here. However, if you expect an exception in some specific lines of your `run` method, please surround them with a try-catch block and catch the specific exception that you expect and act accordingly.

MichaelOchel replied:
Mostly there was NPEs and in this case the run had stoped before all steps was ready. But we hope there will appear no more Exeptions now.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -104,61 +127,101 @@ public Initializer(BridgeHandler bridge, Config config) {
+                if (configChanged) {
+                    updateConfiguration(configuration);
+                }
+                thingUpdatedInt(getThing());
```
Otherwise the properties and so on will not persist. But this is 8 month or more ago, so it may has changed.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -187,76 +250,78 @@ private boolean checkLoginConfig(Config config) {
+                return null;
+            }
+            logger.info("Loading configuration");
+            ArrayList<String> numberExc = null;
```
Because it is only needed to build the error message for the thing status and, if everything is right, the ArrayList is not needed (null). At line 306 the thing status will be build, if it is not null. That the error appears is very improbable, because the parameter has to be a string.

But we can check to an empty list, too, if you think that this is better.

triller-telekom replied:
> But we can check to an empty list, too, if you think that this is better.

Yes, please.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -292,6 +363,12 @@ private Config loadAndCheckConnectionData(Configuration thingConfig) {
+            reconnectTracker.cancel(true);
+        }
+        if (eventListener != null) {
+            eventListener.stop();
```
Because by each little disconnect this method will be called, we only wanted to stop the services and do not delete it. So they can resume their work after a reconnect.

By the way, then the other services would be set to null, too.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/BridgeHandler.java:
```diff
@@ -358,20 +425,28 @@ public synchronized void registerDeviceStatusListener(DeviceStatusListener devic
-                devStatMan.registerDeviceListener(deviceStatusListener);
+                if (devStatMan.getManagerState().equals(ManagerStates.RUNNING)) {
+                    devStatMan.registerDeviceListener(deviceStatusListener);
+                    return;
```
A else-if would be the same. Is that better?

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/DigitalSTROMHandlerFactory.java:
```diff
@@ -115,66 +145,91 @@ protected ThingHandler createHandler(Thing thing) {
-        String dSID;
-        if (StringUtils.isBlank((String) configuration.get(DS_ID))) {
-            dSID = getDSSid(configuration);
+        if (thingUID == null) {
```
We do not know, how it is now, but we think when we implement it 10 or more month ago, it could be called with null.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/discovery/DeviceDiscoveryService.java:
```diff
@@ -88,50 +99,66 @@ protected synchronized void stopScan() {
-                properties.put(DigitalSTROMBindingConstants.DEVICE_UID, device.getDSUID());
+                Map<String, Object> properties = new HashMap<>(1);
                 properties.put(DigitalSTROMBindingConstants.DEVICE_DSID, device.getDSID().getValue());
-                properties.put(DigitalSTROMBindingConstants.DEVICE_HW_INFO, device.getHWinfo());
```
Hm we can delete it, but it was discussed with @kaikreuzer a long time ago.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/event/EventListener.java:
```diff
@@ -0,0 +1,422 @@
+            int index = -1;
+            EventHandler intEventHandler = null;
+            boolean subscribedEventsChanged = false;
+            for (int i = 0; i < eventHandlers.size(); i++) {
```
If no EventHandler is added, the Thread is stopped and, if the first EventHandler was added, the Thread has to be restarted.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/event/EventListener.java:
```diff
@@ -0,0 +1,422 @@
+                eventHandlers.add(eventHandler);
+                addSubscribeEvents(eventHandler.getSupportetEvents());
+                logger.debug("eventHandler: {} added", eventHandler.getUID());
+                if (isStarted) {
```
If no EventHandler is added, the Thread is stopped and, if the first EventHandler was added, the Thread has to be restarted.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java:
```diff
@@ -165,6 +275,10 @@ public void run() {
                     DSID currentDeviceDSID = currentDevice.getDSID();
                     Device eshDevice = tempDeviceMap.remove(currentDeviceDSID);
 
+                    // TODO: Maybe it is better to separate the Device-Status updates (send commands to dSS) and the
```
It is a feature (may be a better solution) we could not implement in the time of our bachelor thesis. Maybe someone else will implement it later.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java:
```diff
@@ -357,38 +526,49 @@ private synchronized void stateChanged(ManagerStates state) {
             pollingScheduler.cancel(true);
             pollingScheduler = null;
-            stateChanged(ManagerStates.STOPPED);
+            logger.debug("stop pollingScheduler");
```
The call is some lines above now.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/serverConnection/impl/DsAPIImpl.java:
```diff
@@ -7,1017 +7,631 @@
 
     @Override
-    public boolean callApartmentScene(String token, int groupID, String groupName, Scene sceneNumber, boolean force) {
+    public boolean callApartmentScene(String token, Short groupID, String groupName, Scene sceneNumber, Boolean force) {
```
Yes, if it is null, it will not be added and the default of digitalstrom will be used.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/serverConnection/impl/DsAPIImpl.java:
```diff
@@ -1157,226 +724,1088 @@ public boolean setDeviceValue(String token, DSID dsid, String name, int value) {
+                                builder.addParameter((String) objAry[0], convertIntegerToString((Integer) objAry[1]));
+                            } else {
+                                builder.buildRequestString();
+                                throw new IllegalArgumentException(
```
No, but this method is currently not used in the binding. We don’t add all functions of the digitalSTROM API, but more than the binding currently need. If something can be set through the binding, too, the methods can be used.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/serverConnection/impl/DsAPIImpl.java:
```diff
@@ -1157,226 +724,1088 @@ public boolean setDeviceValue(String token, DSID dsid, String name, int value) {
+    }
+
+    @Override
+    public JsonObject query2(String sessionToken, String query) {
```
The name is the same like the name of the digitalSTROM function (API) and the difference is explained in the Java-Doc.

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/DetailedGroupInfo.java:
```diff
@@ -13,7 +13,9 @@
- * @author Alexander Betker
+ * @author Alexander Betker - initial contributer
+ * @author Michael Ochel - add java-doc
+ * @author Matthias Siegele - add java-doc
```
We forget it in the PR(s) before. This is the class from OH-1
https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/client/entity/Apartment.java

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/Group.java:
```diff
@@ -10,7 +10,9 @@
- * @author Alexander Betker
+ * @author Alexander Betker - initial contributer
+ * @author Michael Ochel - add java-doc
+ * @author Matthias Siegele - add java-doc
```
same:
https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/client/entity/Group.java

MichaelOchel commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/Apartment.java:
```diff
@@ -12,7 +12,9 @@
- * @author Alexander Betker
+ * @author Alexander Betker - initial contributer
+ * @author Michael Ochel - add java-doc
+ * @author Matthias Siegele - add java-doc
```
We forget it in the PR(s) before. This is the class from OH-1
https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/client/entity/Apartment.java

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java:
```diff
@@ -672,7 +668,7 @@ public synchronized void sendSceneComandsToDSS(InternalScene scene, boolean call
                     Thread.sleep(sleepTime);
                 } catch (InterruptedException e) {
-                    logger.error("An exception occurred", e);
+                    logger.error("An InterruptedException occurred", e);
```
Since this will end up in a user's log file, is this really something that he should report? Or is it a rather rare exception that occurs while testing/debugging? In the later case I would vote to reduce the severity to `debug` here.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/CircuitHandler.java:
```diff
@@ -108,14 +109,14 @@ private synchronized BridgeHandler getDssBridgeHandler() {
             if (handler instanceof BridgeHandler) {
                 dssBridgeHandler = (BridgeHandler) handler;
-                dssBridgeHandler.registerDeviceStatusListener(this);
+                // dssBridgeHandler.registerDeviceStatusListener(this);
```
maybe its better to fully remove these lines instead of just commenting them out.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/CircuitHandler.java:
```diff
@@ -108,14 +109,14 @@ private synchronized BridgeHandler getDssBridgeHandler() {
-        if (circuit == null) {
-            dssBridgeHandler.registerDeviceStatusListener(this);
-        }
+        // if (circuit == null) {
```
maybe its better to fully remove these lines instead of just commenting them out.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/DeviceHandler.java:
```diff
@@ -283,15 +289,15 @@ private synchronized BridgeHandler getDssBridgeHandler() {
-                if (device == null) {
-                    dssBridgeHandler.registerDeviceStatusListener(this);
-                }
+                // if (device == null) {
```
maybe its better to fully remove these lines instead of just commenting them out.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/DeviceHandler.java:
```diff
@@ -283,15 +289,15 @@ private synchronized BridgeHandler getDssBridgeHandler() {
-        } else if (device == null) {
-            dssBridgeHandler.registerDeviceStatusListener(this);
-        }
+        } // else if (device == null) {
```
maybe its better to fully remove these lines instead of just commenting them out.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/ZoneTemperatureControlHandler.java:
```diff
@@ -259,9 +260,9 @@ private synchronized BridgeHandler getDssBridgeHandler() {
-        if (temperatureSensorTransmitter == null && zoneID != null) {
-            dssBridgeHandler.registerTemperatureControlStatusListener(this);
-        }
+        // if (temperatureSensorTransmitter == null && zoneID != null) {
```
maybe its better to fully remove these lines instead of just commenting them out.

hmerk commented:
@triller-telekom of course, I will be away for the weekend, returning Tuesday, so will give the new version a „stress-test“ when I am back.

triller-telekom commented:
@MichaelOchel Thank you for the changes. Can you please also do a rebase of all your commits onto the current master? If a test from @hmerk is successful next week we can then start a CQ at Eclipse which will do a legal check on the code before we are allowed to merge it, since it contains more than 1000 LOC.

MichaelOchel commented:
I'm at it right now. 

Should i also squash it?

triller-telekom commented:
No, please do not squash it. Squashing will be automatically done once we merge it. Thanks.

MichaelOchel commented:
Ok, I was at the last step, so I pushed it first without without squashing. 

... Ok, than I have done all right ^^

hmerk commented:
Hi,
I wanted to start testing, but I cannot pull the actual changes into my local repo cause of some conflicts.
@MichaelOchel can you provide me your latest binding jar for testing.

kaikreuzer commented:
@hmerk The conflicts were introduced to a recent refactoring on the MDNS discovery - I have just rebased this PR here myself in order to resolve the conflicts.

Please note that you can only test this code on the latest ESH - it won't work with the openHAB snapshot; I plan to do a new distro build tomorrow, which should then be compatible again, so you might want to wait for it.

triller-telekom commented:
@hmerk I do not want to rush you, but since Kai has released the new ESH stable last friday, did you have a chance to test this PR, yet?

hmerk commented:
@triller-telekom 
Hi Stefan, unfortunately not, but will give it a try this evening.

hmerk commented:
@triller-telekom 
I just did a fresh clone of eclipse/smarthome and get the following error on mvn

[ERROR] Failed to resolve target definition /home/hmerk/.m2/repository/org/eclipse/smarthome/targetplatform/0.9.0-SNAPSHOT/targetplatform-0.9.0-SNAPSHOT-smarthome.target: Could not find "org.eclipse.equinox.sdk.feature.group/3.13.0.v20170531-1133" in the repositories of the current location -> [Help 1]

hmerk commented:
@MichaelOchel @triller-telekom 
Got it running in a fresh dev-env, here is my first impression:
Started run debug configuration -> No autodiscovery off DSS, inbox empty.
Added mannually -> discovery started, DSS found, after adding applicationToken, DSS went online and my devices where discovered and shown in inbox.
Controlling a SW-ZWS200, GE-TKM210 and SW-KL200 through PaperUI was successfull. Did not try my rollershutters cause wife already has gone to bed ....
--

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/OSGI-INF/DigitalSTROMBridgeMDNSDiscoveryParticipant.xml:
```diff
@@ -1,17 +1,17 @@
+	name="org.eclipse.smarthome.binding.digitalstrom.discovery.DigitalSTROMMDNSDiscoveryParticipant">
+	<implementation class="org.eclipse.smarthome.binding.digitalstrom.internal.discovery.BridgeMDNSDiscoveryParticipant" />
+	<service>
+  <provide interface="org.eclipse.smarthome.config.discovery.mdns.MDNSDiscoveryParticipant"/>
```
could you please re-run the auto-formatter here?

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/OSGI-INF/DigitalSTROMDeviceTypeProvider.xml:
```diff
@@ -0,0 +1,18 @@
+	name="org.eclipse.smarthome.binding.digitalstrom.internal.providers.DsDeviceThingTypeProvider">
+	<implementation class="org.eclipse.smarthome.binding.digitalstrom.internal.providers.DsDeviceThingTypeProvider" />
+    <reference bind="setTranslationProvider" cardinality="1..1" interface="org.eclipse.smarthome.core.i18n.TranslationProvider" name="TranslationProvider" policy="dynamic" unbind="unsetTranslationProvider"/>
+ <service>
```
could you please re-run the auto-formatter here?

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/handler/DeviceHandler.java:
```diff
@@ -300,120 +289,121 @@ private synchronized BridgeHandler getDssBridgeHandler() {
-                                        updateState(new ChannelUID(getThing().getUID(), currentChannel), OnOffType.ON);
+                                } else if (currentChannel.contains(DsChannelTypeProvider.STAGE)) {
+                                    if (currentChannel.contains(TWO_STAGE_SWITCH_IDENTICATOR)) {
+                                        updateState(new ChannelUID(getThing().getUID(), currentChannel),
```
Could you replace all those occurrences in this class from channelUID to channelID. Here it would simply be:
```
updateState(currentChannel, new StringType(convertStageValue((short) 2, device.getOutputValue())));
```

MichaelOchel commented:
Thanks for your help, too.

We incorporated your last comments and squashed our commits. We hope it is all fine now and it can be merged soon. :)

Best regards

Matthias and Michael

sjka commented:
FTR: I filed this as CQ 15035

maggu2810 commented:
AFAIK CQ 15035 has been approved. So, ready to merge?

kaikreuzer commented:
yes, but we should add a second commit to the branch, which adapts the license headers.

kaikreuzer commented:
I have added a commit, which updates the license headers and fixes the SAT errors.

Once Travis succeeds, this can be merged - **please merge by rebase, not by squash**.

sjka commented:
Big PR -> big thanks to everybody involved!
